external/fv3fit/fv3fit/train_microphysics.py

```diff
@@ -386,6 +391,26 @@ def open_dataset(
             random_state=np.random.RandomState(0),
         )
 
+    def open_dataset(
+        self, url: str, nfiles: Optional[int], required_variables: Set[str],
+    ) -> tf.data.Dataset:
+        if self.data_format == "nc":
+            self._open_dataset(
+                url, nfiles, required_variables,
+            )
+        elif self.data_format == "tf":
+            # needs to be batched
+            return tf.data.experimental.load(url).batch(1)
+        else:
+            raise NotImplementedError(self.data_format)
+
+    def open_train_test(self):
```

Comment:
I think there is currently a bug where if you don't call the `save` workflow and train on its output, the `train` workflow never applies the filters even though it says it does. It would also easy to accidentally introduce a bug in the save-then-train workflow by calling this function in `train_main`, since that would mean filters get applied twice.

If you separate "save" into its own script you could remove the filters entirely from the train workflow, which I think solves the bug.

There are different ways to make it clear open_train_test should only be called when you want to filter, one option is to move this into the file for `save` and made a private function.

Reply:
This is the only really important comment in this review.

Reply:
> since that would mean filters get applied twice.

What is wrong with applying filters twice? The result is the same as applying once.

>  think there is currently a bug where if you don't call the save workflow and train on its output, the train workflow never applies the filters even though it says it does

Perhaps I am missing something. Could you say more? As far as I can tell all the CLI entrypoints will apply the filter.

1. fv3fit.train_microphysics -> train_main -> train_function (which calls `prepare_flat_data`) -> _train_function_unbatched
1. fv3fit.train -> train_function (which calls prepare_flat_data)
1. save -> open_train_test (which calls prepare_flat_data).

Reply:
I hadn't realized what the filters do, but given it's filtering on variables it should be fine to apply them twice.

I missed the call to `prepare_flat_data` in `train_function`.

Is there any risk to the other parts of `prepare_flat_data` being called twice? Particularly `.unbatch()`? If not the PR should be bug-free, though it could be easy for someone editing this to introduce a bug.

Reply:
I agree it's a concern, but I feel like the signature and name of `prepare_flat_data` makes it clear that the output data is unbatched while the input is batched.

Reply:
Looking longer, I was missing the extra `.batch()` call that happens when loading from tfdataset which makes it fine to call `.unbatch()` twice.

Reply:
I wish we could strongly type dataset dimensions.